### PR TITLE
Endpoints for plio's, user's and org's settings

### LINKED ISSUE #281: A way to update/create the `settings` key inside the config (JSONField) of Plio, User and Organizations models

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #281 
Requires [this](https://github.com/avantifellows/plio-frontend/pull/570) to be merged

## Test Plan

- [x] Wrote tests
- [x] Tested locally
- [ ] Tested on staging
- [ ] Tested on production
- [ ] If changes in DB, update [DB schema](https://docs.google.com/presentation/d/15c45HbHphh4w3guaFV3BgEA5nMKd-4J35JdUKdrpPzo/edit#slide=id.gc8c9ec0b5e_0_0) and BigQuery (staging and prod)

### CODE REVIEW

File: plio/tests.py [hunk 1]
```diff
@@ -502,13 +502,18 @@ def test_items_sorted_with_time(self):
         self.assertEqual(response.data["items"][1]["id"], item_1.id)
 
     def test_retrieving_plio_sets_instance_cache(self):
+        # create a third plio
+        plio_3 = Plio.objects.create(
+            name="Plio 3", video=self.video, created_by=self.user
+        )
+
```
Comment (dalmia) [inline plio/tests.py:509]: any reason why a new plio was created for this?

File: plio/tests.py [hunk 2]
```diff
@@ -541,6 +546,99 @@ def test_updating_plio_recreates_instance_cache(self):
...
         self.assertEqual(cache.get(cache_key_name)["name"], new_name)
 
+    def test_user_can_update_own_plio_settings(self):
+        test_settings = json.dumps(
+            {"player": {"configuration": {"skipEnabled": False}}}
+        )
+        # update settings for plio_1
+        response = self.client.put(
+            f"/api/v1/plios/{self.plio_1.uuid}/setting/",
+            json.loads(test_settings),
```
Comment (dalmia) [inline plio/tests.py:556]: `json.loads(json.dumps(X)) = X` right?

**Review by dalmia — CHANGES_REQUESTED** [review]

File: users/tests.py [hunk 3]
```diff
@@ -362,6 +362,63 @@ def test_updating_user_recreates_instance_cache(self):
...
+        # update settings for user 1
+        response = self.client.put(
+            f"/api/v1/users/{self.user.id}/setting/",
+            json.loads(test_settings),
+            format="json",
+        )
+
+        # 200 OK returned as status
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # The plio should contain the new updated settings object
```
Comment (dalmia) [inline users/tests.py:381]: we have generally not including capital letters in general comment sentences
```suggestion
        # the plio should contain the new updated settings object
```

File: users/tests.py [hunk 4]
```diff
@@ -362,6 +362,63 @@ def test_updating_user_recreates_instance_cache(self):
...
+                "player": {"configuration": {"skipEnabled": False}},
+                "app": {"appearance": {"darkMode": True}},
+            }
+        )
+        # try to update settings for user_2
+        response = self.client.put(
+            f"/api/v1/users/{self.user_2.id}/setting/",
+            json.loads(test_settings),
+            format="json",
+        )
```
Comment (dalmia) [inline users/tests.py:399]: same point about json.dumps and json.loads being redundant

File: users/tests.py [hunk 5]
```diff
@@ -362,6 +362,63 @@ def test_updating_user_recreates_instance_cache(self):
...
+
+        # Forbidden action
```
Comment (dalmia) [inline users/tests.py:401]:
```suggestion
        # forbidden action
```

File: users/tests.py [hunk 6]
```diff
@@ -362,6 +362,63 @@ def test_updating_user_recreates_instance_cache(self):
...
+        self.user.save()
+
+        # try the request again
+        response = self.client.put(
+            f"/api/v1/users/{self.user_2.id}/setting/",
+            json.loads(test_settings),
+            format="json",
+        )
+
+        # 200 OK returned as status
```
Comment (dalmia) [inline users/tests.py:415]: i don't think we need to write comments for every single action/line - this is fairly obvious - let's remove it?

File: users/tests.py [hunk 7]
```diff
@@ -362,6 +362,63 @@ def test_updating_user_recreates_instance_cache(self):
...
+        response = self.client.put(
+            f"/api/v1/users/{self.user_2.id}/setting/",
+            json.loads(test_settings),
+            format="json",
+        )
+
+        # Forbidden action
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+        # make user 1 the superuser
```
Comment (dalmia) [inline users/tests.py:404]: i don't think we need to write comments for every single action/line - this is fairly obvious - let's remove it?

File: users/tests.py [hunk 8]
```diff
@@ -362,6 +362,63 @@ def test_updating_user_recreates_instance_cache(self):
...
+        self.user.is_superuser = True
+        self.user.save()
+
+        # try the request again
```
Comment (dalmia) [inline users/tests.py:408]: i don't think we need to write comments for every single action/line - this is fairly obvious - let's remove it?

File: plio/tests.py [hunk 9]
```diff
@@ -503,13 +503,18 @@ def test_items_sorted_with_time(self):
         self.assertEqual(response.data["items"][1]["id"], item_1.id)
 
     def test_retrieving_plio_sets_instance_cache(self):
+        # create a third plio
+        plio_3 = Plio.objects.create(
+            name="Plio 3", video=self.video, created_by=self.user
+        )
```
Comment (dalmia) [inline plio/tests.py:509]: any reason why a new plio was created for this?

File: plio/tests.py [hunk 10]
```diff
@@ -542,6 +547,99 @@ def test_updating_plio_recreates_instance_cache(self):
...
+    def test_user_can_update_own_plio_settings(self):
+        test_settings = json.dumps(
+            {"player": {"configuration": {"skipEnabled": False}}}
+        )
+        # update settings for plio_1
+        response = self.client.put(
+            f"/api/v1/plios/{self.plio_1.uuid}/setting/",
+            json.loads(test_settings),
+            format="json",
+        )
```
Comment (dalmia) [inline plio/tests.py:559]: same point about json.dumps and json.loads being redundant?

File: plio/tests.py [hunk 11]
```diff
@@ -542,6 +547,99 @@ def test_updating_plio_recreates_instance_cache(self):
...
+
+    def test_user_cannot_update_other_user_plio_settings(self):
+        test_settings = json.dumps(
+            {"player": {"configuration": {"skipEnabled": False}}}
+        )
+        # user_2 creates a plio
+        plio = Plio.objects.create(
+            name="Plio_by_user_2", video=self.video, created_by=self.user_2
+        )
+
```
Comment (dalmia) [inline plio/tests.py:579]: same point about json.dumps and json.loads being redundant?

File: plio/tests.py [hunk 12]
```diff
@@ -542,6 +547,99 @@ def test_updating_plio_recreates_instance_cache(self):
...
+        # user_1 tries updating the settings for the above created plio
+        response = self.client.put(
+            f"/api/v1/plios/{plio.uuid}/setting/",
+            json.loads(test_settings),
+            format="json",
+        )
+
+        # Updating other user's plio's settings should be forbidden
```
Comment (dalmia) [inline plio/tests.py:587]:
```suggestion
        # updating other user's plio's settings should be forbidden
```

File: plio/tests.py [hunk 13]
```diff
@@ -542,6 +547,99 @@ def test_updating_plio_recreates_instance_cache(self):
...
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_org_members_can_update_org_plio_settings(self):
+        test_settings = json.dumps(
+            {"player": {"configuration": {"skipEnabled": False}}}
+        )
```
Comment (dalmia) [inline plio/tests.py:593]: same point about json.dumps and json.loads being redundant?

File: organizations/tests.py [hunk 14]
```diff
@@ -91,3 +91,130 @@ def test_updating_organization_recreates_user_instance_cache(self):
         self.assertEqual(
             cache.get(cache_key_name)["organizations"][0]["name"], org_new_name
         )
+
+    def test_settings_support_only_put_method(self):
+        """ /settings/ action only supports PUT method """
```
Comment (dalmia) [inline organizations/tests.py:96]: i think we have started the right thing of not unnecessarily writing docstrings for tests when the function name makes it obvious. so, let's remove this?

File: organizations/tests.py [hunk 15]
```diff
@@ -91,3 +91,130 @@ def test_updating_organization_recreates_user_instance_cache(self):
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(
+            Organization.objects.filter(id=self.organization_1.id)
+            .first()
+            .config["settings"],
+            dummy_settings,
+        )
+
+    def test_superuser_can_update_any_org_settings(self):
+        """A superuse is allowed to update settings of any org"""
```
Comment (dalmia) [inline organizations/tests.py:135]: same point about redundant docstring in tests

File: organizations/tests.py [hunk 16]
```diff
@@ -91,3 +91,130 @@ def test_updating_organization_recreates_user_instance_cache(self):
...
+        response = self.client.put(
+            f"/api/v1/organizations/{self.organization_2.id}/setting/", dummy_settings
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(
+            Organization.objects.filter(id=self.organization_2.id)
+            .first()
+            .config["settings"],
+            dummy_settings,
+        )
```
Comment (dalmia) [inline organizations/tests.py:165]: what extra is this checking?

File: organizations/tests.py [hunk 17]
```diff
@@ -91,3 +91,130 @@ def test_updating_organization_recreates_user_instance_cache(self):
...
+
+        # user should NOT be able to update org 1 settings as
+        # the user is not an org admin
+        response = self.client.put(
+            f"/api/v1/organizations/{self.organization_1.id}/setting/", dummy_settings
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+        # create a new super user
+        superuser_client = APIClient()
```
Comment (dalmia) [inline organizations/tests.py:184]: why not use self.client?

**Review by dalmia — CHANGES_REQUESTED** [review]

File: organizations/views.py [hunk 18]
```diff
@@ -20,3 +22,19 @@ class OrganizationViewSet(viewsets.ModelViewSet):
     permission_classes = [IsAuthenticated, OrganizationPermission]
     queryset = Organization.objects.all()
     serializer_class = OrganizationSerializer
+
+    @action(
+        detail=True,
+        permission_classes=[IsAuthenticated, OrganizationPermission],
+        methods=["put"],
+    )
+    def setting(self, request, pk):
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply organizations/views.py:31]: https://github.com/encode/django-rest-framework/issues/3337

`settings` is a keyword reserved for Django so we cannot use that name in a view

File: plio/views.py [hunk 19]
```diff
@@ -150,6 +150,22 @@ def perform_create(self, serializer):
     def perform_update(self, serializer):
         serializer.save(created_by=self.get_object().created_by)
 
+    @action(
+        detail=True,
+        permission_classes=[IsAuthenticated, PlioPermission],
+        methods=["put"],
+    )
+    def setting(self, request, uuid):
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply plio/views.py:157]: same comment as above

File: users/views.py [hunk 20]
```diff
@@ -45,6 +45,23 @@ class UserViewSet(viewsets.ModelViewSet):
     queryset = User.objects.all()
     serializer_class = UserSerializer
 
+    @action(
+        detail=True,
+        permission_classes=[IsAuthenticated, UserPermission],
+        methods=["put"],
+    )
+    def setting(self, request, pk):
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply users/views.py:51]: same comment as above

File: organizations/views.py [hunk 21]
```diff
@@ -26,12 +26,12 @@ class OrganizationViewSet(viewsets.ModelViewSet):
     @action(
         detail=True,
         permission_classes=[IsAuthenticated, OrganizationPermission],
-        methods=["put"],
+        methods=["patch"],
     )
     def setting(self, request, pk):
-        """Updates an org's setting key inside it's config"""
+        """Updates an org's settings key inside it's config"""
```
Comment (dalmia) [inline organizations/views.py:32]: similar to what we did for plio

```suggestion
        """Updates an org's settings"""
```

File: users/views.py [hunk 22]
```diff
@@ -46,14 +46,11 @@ class UserViewSet(viewsets.ModelViewSet):
     @action(
         detail=True,
         permission_classes=[IsAuthenticated, UserPermission],
-        methods=["put"],
+        methods=["patch"],
     )
     def setting(self, request, pk):
         """Updates a user's setting key inside it's config"""
```
Comment (dalmia) [inline users/views.py:52]:
```suggestion
        """Updates a user's settings"""
```

File: organizations/tests.py (hunk 17 above)
Comment (dalmia) [reply organizations/tests.py:182]: do drop a comment about this

**Review by dalmia — CHANGES_REQUESTED** [review]

File: organizations/views.py [hunk 23]
```diff
@@ -20,3 +22,19 @@ class OrganizationViewSet(viewsets.ModelViewSet):
...
+    )
+    def setting(self, request, pk):
+        """Updates an org's settings"""
+        org = self.get_object()
+        config = org.config if org.config is not None else {}
+        config["settings"] = self.request.data
+        org.config = config
+        org.save()
+        return Response(
+            self.get_serializer(org).data["config"], status=status.HTTP_200_OK
```
Comment (deepansh96) [inline organizations/views.py:39]:
```suggestion
            self.get_serializer(org).data["config"]
```

**Review by dalmia — APPROVED** [review]

**Pull request merged**